lib/reply.js

```diff
@@ -140,27 +141,36 @@ function wrapOnSendEnd (err, reply, payload) {
 }
 
 function onSendEnd (reply, payload) {
+  const res = reply.res
+
   if (payload === undefined || payload === null) {
     reply.sent = true
-    reply.res.end()
+    reply._headers['content-length'] = '0'
+    res.writeHead(res.statusCode, reply._headers)
+    // avoid ArgumentsAdaptorTrampoline from V8
+    res.end(null, null, null)
     return
   }
 
   if (typeof payload.pipe === 'function') {
-    sendStream(payload, reply.res, reply)
+    sendStream(payload, res, reply)
     return
   }
 
   if (typeof payload !== 'string' && !Buffer.isBuffer(payload)) {
     throw new TypeError(`Attempted to send payload of invalid type '${typeof payload}'. Expected a string or Buffer.`)
   }
 
-  if (!reply.res.getHeader('Content-Length')) {
-    reply.res.setHeader('Content-Length', '' + Buffer.byteLength(payload))
+  if (!reply._headers['content-length']) {
```

Comment:
Same thing as my other comment about the header possibly being set on `res`.

Reply:
This is the final handler of the `onSend` hooks, so if a user wants to alter the headers it will use the `reply` interface. I hope that no one sets the `content-length` header inside a `onRequest` or `preHandler` hook.

Reply:
I also hope that no one will do that, but we can't stop them from doing it so it's a case that needs to be handled.

Reply:
I don't agree, I don't think we should drop a considerable amount of ops/sec just because a user **may** do something that is wrong.